File: scraper/db_manager.py
import sqlite3
import os
from datetime import datetime, timedelta

class DBManager:
    def __init__(self, db_path="jobs_state.db"):
        self.db_path = db_path
        self._init_db()

    def _init_db(self):
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS processed_jobs (
                    url TEXT PRIMARY KEY,
                    last_scraped DATETIME,
                    hash TEXT,
                    status TEXT
                )
            """)
            conn.commit()

    def is_processed(self, url, content_hash=None):
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute("SELECT hash FROM processed_jobs WHERE url = ?", (url,))
            row = cursor.fetchone()
            if not row:
                return False
            if content_hash and row[0] != content_hash:
                return False
            return True

    def save_job(self, url, content_hash, status="inserted"):
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("""
                INSERT OR REPLACE INTO processed_jobs (url, last_scraped, hash, status)
                VALUES (?, ?, ?, ?)
            """, (url, datetime.now().isoformat(), content_hash, status))
            conn.commit()

    def prune_old_records(self, days=365):
        cutoff = (datetime.now() - timedelta(days=days)).isoformat()
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("DELETE FROM processed_jobs WHERE last_scraped < ?", (cutoff,))
            conn.commit()

File: scraper/models/db_manager.js
const sqlite3 = require('sqlite3').verbose();
const path = require('path');

class DBManager {
    constructor() {
        this.dbPath = path.resolve(__dirname, '../scraper.db');
        this.db = new sqlite3.Database(this.dbPath);
    }

    async init() {
        return new Promise((resolve, reject) => {
            this.db.serialize(() => {
                this.db.run(`
                    CREATE TABLE IF NOT EXISTS jobs (
                        url TEXT PRIMARY KEY,
                        title TEXT,
                        last_scraped_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                        hash TEXT
                    )
                `, (err) => {
                    if (err) reject(err);
                    else resolve();
                });
            });
        });
    }

    async isDuplicate(url) {
        return new Promise((resolve, reject) => {
            this.db.get('SELECT url FROM jobs WHERE url = ?', [url], (err, row) => {
                if (err) reject(err);
                else resolve(!!row);
            });
        });
    }

    async saveJob(url, title, hash) {
        return new Promise((resolve, reject) => {
            this.db.run(
                'INSERT OR REPLACE INTO jobs (url, title, hash, last_scraped_at) VALUES (?, ?, ?, CURRENT_TIMESTAMP)',
                [url, title, hash],
                (err) => {
                    if (err) reject(err);
                    else resolve();
                }
            );
        });
    }

    async pruneOldRecords() {
        return new Promise((resolve, reject) => {
            // Delete records older than 1 year (365 days)
            this.db.run(
                "DELETE FROM jobs WHERE last_scraped_at < date('now', '-365 days')",
                (err) => {
                    if (err) reject(err);
                    else {
                        console.log('Old records pruned from local database.');
                        resolve();
                    }
                }
            );
        });
    }

    close() {
        this.db.close();
    }
}

module.exports = new DBManager();
